example/composer.json

```
@@ -4,12 +4,9 @@
         "php": "~8.2.0 || ~8.3.0 || ~8.4.0 || ~8.5.0",
         "illuminate/container": "^12.40.2",
         "laminas/laminas-servicemanager": "^4.5.0",
-        "symfony/cache": "^7.4.0"
+        "symfony/cache": "^8.0.0"
     },
     "config": {
-        "platform": {
-            "php": "8.2.99"
-        },
```

This should stay to prevent out-of-version Renovate `composer.lock` changes

---

So, I have no idea what you are talking about, but what I know for sure is `config.platform` kills the purpose of a multi-environment CI build. Let me explain with this hypotetical `foo/bar` library:

- PHPTAL `v1.11.0` supports `~8.2.0 || ~8.3.0 || ~8.4.0`
- PHPTAL `v1.12.0` supports `~8.3.0 || ~8.4.0 || ~8.5.0`

```console
# php -v
PHP 8.5.0 (cli)

# cat composer.json 
{
    "name": "foo/bar",
    "require": {
        "phptal/phptal": "~1.10.0 || ~1.11.0"
    },
    "config": {
        "platform": {
            "php": "8.2.99"
        }
    }
}

# composer show
phptal/phptal 1.10.0
```

As you can see, even though `PHPTAL:~1.11.0` is supported **and consumers of `foo/bar` install `PHPTAL:~1.11.0`**, we would never test our library against it!

That should be considered completely unacceptable, and I'm clueless on why you keep pushing for it.

---

I made it very clear: `config.platform` locks the `composer.lock` upgrades, or else Renovate will just do a mess.

What you can do in `@lowest` and `@latest` is dropping it programmatically, but you need a locked run anyway, and that one depends heavily on `config.platform`

---

Following https://github.com/lcobucci/jwt/pull/1106

> From my PoV, reproducible builds are way more important than being able to try latest/greatest

This is something I foundamentally disagree on.

Reproducible builds and `lock` files are crucial for a final product confined in an environment I can control and take questionable decisions on: "_One of our components has a security flaw, but we have enough active mitigations to take care of something else first, like data integrity and business protocols_"

On libraries instead you have to assume every consumer is already on the latest dep because of security updates and infrastructure evolution.
You have a green build on your `lock` file of 6 months ago: so what? Who the heck cares? We moved on, we need a green build for today's software, we can't deploy the 6-month old anymore as much as you do in your private production facility.

---

> Who the heck cares?

I care, **deeply**, because I must be able to check out the project, bisect the change that caused a failure, or just go back to it and do meaningful work.

The audience here is the maintainer of the project (me).

In fact, I just did a ton of PHP 8.5 upgrades (15~20 repositories, over the last 3 months), and the only reason why I could confidently perform them, was because I was able to install the exact dependency set as for the builds that failed. Yes, they were stuck with an old `.lock` for many months, and that actually **helped** with getting them up to speed again.

The lock file is a clear requirement for meaningful project collaboration, even at the cost of restricted ranges for downstream (due to our testing forbidding `symfony/*:^8`).

Even dropping the lock file doesn't help: you get dependency range bumps that aren't compatible with your system dependencies. At least with the `.lock` file you see immediately that the bump failed due to `.lock` not being re-generated correctly (renovate tells you that).

---

It bothers me a lot how lightly you passed over the fact declared dependencies are silently ignored and never tested on.

I won't talk further about `lock` files for libraries: we disagree as I see them as burden, but I can live with them.

`config.platform` on the other hand is dangerous.

> What you can do in `@lowest` and `@latest` is dropping it programmatically

A configuration that I have to remember to drop at runtime to be sure I test everything correctly?
Don't you see how bad it is?

> This should stay to prevent out-of-version Renovate `composer.lock` changes

Using `config.platform` as a solution to this problem is wrong: if you want a `lock` file, then the CI should run `locked` only on the very specific environment the `lock` file was created on, not the entire matrix.

I'm shocked when I think of all the repositories of yours that have this bug, and the amount of officially supported but never tested dependencies that currently run in the wild.

---

> A configuration that I have to remember to drop at runtime to be sure I test everything correctly?
> Don't you see how bad it is?

No, I don't: it's fine to do this in CI, and you can propose it for the upstream laminas CI tooling too.

---

> No, I don't

Well, maybe your sight needs a checkup then.

A `lock` file is a snapshot of the state of a system, that allows you to reproduce the system and thus the build.
Currently it seems to me `composer` doesn't store the PHP version the `lock` file was generated on, and `renovatebot` isn't particularly clear on what PHP version is using to regenerate the `lock` file: what a pity.

But one this has to be sure: `renovatebot` is using only **one** version of PHP to recreate it.
Then: what's the point of running a reproducible build in an environment different from the one the `lock` file has been generated on?

You are running `locked` dependencies on multiple envs, and then you add `config.platform` to not get annoyed by `renovatebot`: `config.platform` is the wrong solution to an actual problem, where the correct solution is running `locked` only on the same env `renovatebot` used to recreated the `lock` file.

With `config.platform`, at best you don't care about testing your declared dependencies, at worst you like to lie to your consumers about the actual compatibilities of your library.

The new push reflects this comment.

---

> and `renovatebot` isn't particularly clear on what PHP version is using to regenerate the `lock` file

It uses `config.platform.php`, hence why it was introduced. I spent a good couple weeks around that problem, at the time, and this was the most clean/declarative solution, while other solutions affected downstream.

> But one this has to be sure: `renovatebot` is using only **one** version of PHP to recreate it.
> Then: what's the point 

Correct. The point is being able to pick up the pieces from that exact version, if anything breaks.

> With `config.platform`, at best you don't care about testing your declared dependencies, at worst you like to lie to your consumers about the actual compatibilities of your library.

As I said, you can declaratively drop it before running `@lowest` or `@latest`, and it is a good idea to do this in upstream pipeline abstractions.

EDIT: to be clear, there's a `composer` command for it, IIRC (and worst case a `| jq` transformation)

---

> I spent a good couple weeks around that problem, at the time, and this was the most clean/declarative solution, while other solutions affected downstream.

I have no evidence that this is true.

The solution I'm proposing here cares about `locked`, `lowest` and `highest` by default and by design without any additional step to be remembered and maintained, and declares as well which PHP version we care `locked` file should run on.

---

With your changes applied, the next run of Renovate update `composer.lock` for what runs on PHP 8.5, not what runs on PHP 8.4.

As for the "good couple weeks" I mentioned, we went back and forth with Renovate on this: https://github.com/renovatebot/renovate/issues/18715 is the main issue, and there's a bunch of satellite issues for which I don't have clear recollection (started taking detailed personal notes a year later)

---

More refs:

* https://github.com/laminas/technical-steering-committee/blob/9fef4752cc235b30ed79094e013c802a2db5ebc1/meetings/minutes/2022-02-07-TSC-Minutes.md?plain=1#L36
* https://github.com/laminas/technical-steering-committee/commit/6f11ab7ed5c96fc524424c3591e0b5236c7f522e

---

Turns out that the CI matrix action already does exactly what I was asking here:

* https://github.com/laminas/laminas-continuous-integration-action/blob/a4d9b3138e22fd99185b0c05f2cfdf4158e821ce/entrypoint.sh#L103-L106
* https://github.com/laminas/laminas-continuous-integration-action/commit/dc5580f20107ec4414347f02fcbdd3f549d7291a
* https://github.com/laminas/laminas-continuous-integration-action/pull/79